internal/report/snapshot/snapshot_buffers.go

```diff
@@ -0,0 +1,255 @@
+		oldest := l.buffer[0]
+		newBufferSize := logSize + bufferSize
+		if newBufferSize-oldest.LogRecordCount() >= l.idealSize {
+			l.buffer = l.buffer[1:]
+		}
+	}
+}
+
+// ConstructPayload condenses the buffer and serializes to protobuf
+func (l *LogBuffer) ConstructPayload() ([]byte, error) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	logsMarshaler := plog.NewProtoMarshaler()
+
+	payloadLogs := plog.NewLogs()
+	for _, ld := range l.buffer {
+		ld.ResourceLogs().MoveAndAppendTo(payloadLogs.ResourceLogs())
+	}
+
+	// update the buffer to retain the current logs which were moved to the new payload
+	l.buffer = []plog.Logs{payloadLogs}
+
+	payload, err := logsMarshaler.MarshalLogs(payloadLogs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct payload: %w", err)
+	}
+
+	return payload, nil
+}
+
+// MetricBuffer is a buffer for pmetric.Metrics
+type MetricBuffer struct {
+	mutex     sync.Mutex
+	buffer    []pmetric.Metrics
+	idealSize int
+}
+
+// NewMetricBuffer creates a metricBuffer with the ideal size set
+func NewMetricBuffer(idealSize int) *MetricBuffer {
+	return &MetricBuffer{
+		buffer:    make([]pmetric.Metrics, 0),
+		idealSize: idealSize,
+	}
+}
+
+// Len counts the number of data points in all Metric payloads in buffer
+func (l *MetricBuffer) Len() int {
+	size := 0
+	for _, md := range l.buffer {
+		size += md.DataPointCount()
+	}
+
+	return size
+}
+
+// Add adds the new metric payload and adjust buffer to keep ideal size
+func (l *MetricBuffer) Add(md pmetric.Metrics) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	metricSize := md.DataPointCount()
+	bufferSize := l.Len()
+	switch {
+	// The number of metrics is more than idealSize so reset this to just this metric set
+	case metricSize > l.idealSize:
+		l.buffer = []pmetric.Metrics{md}
+
+	// Haven't reached idealSize yet so add this
+	case metricSize+bufferSize < l.idealSize:
+		l.buffer = append(l.buffer, md)
+
+	// Adding this will put us over idealSize so and add the new metrics.
+	// Only remove the oldest if it does not bring buffer under idealSize
+	case metricSize+bufferSize >= l.idealSize:
+		l.buffer = append(l.buffer, md)
+
+		// If removing the oldest item does not put us under the ideal size then it's ok to do so
+		oldest := l.buffer[0]
+		newBufferSize := metricSize + bufferSize
+		if newBufferSize-oldest.DataPointCount() >= l.idealSize {
+			l.buffer = l.buffer[1:]
+		}
+	}
+}
+
+// ConstructPayload condenses the buffer and serializes to protobuf
+func (l *MetricBuffer) ConstructPayload() ([]byte, error) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	metricMarshaler := pmetric.NewProtoMarshaler()
+
+	payloadMetrics := pmetric.NewMetrics()
+	for _, md := range l.buffer {
+		md.ResourceMetrics().MoveAndAppendTo(payloadMetrics.ResourceMetrics())
+	}
+
+	// update the buffer to retain the current metrics which were moved to the new payload
+	l.buffer = []pmetric.Metrics{payloadMetrics}
```

Comment:
Something I find noteworthy, this will make it so that we tend to remove these payloads after a report, since we always drop the whole batch when removing from the buffer when it gets full. Seems fine, just interesting.

Reply:
This logic was because we move all the items from the buffer into the payload. We don't want to lose them incase we get another report call right away so we just overwrite the buffer with its old coalesced contents. These will stick around until there are enough items added that removing this won't put us under the `idealSize`